### ISSUE TITLE
SwapBuffers, GL.Flush, or GL.Finish Takes a While After a Compute Shader

### ISSUE THREAD

**Issue by Metric** [opentk/opentk#985]
### Description

SwapBuffers, GL.Flush, or GL.Finish takes a while to finish when a compute shader has been used, even though the compute shader is already finished on Win10 NVIDIA 1080 with Drivers: 441.41. Thus, causing the UI thread to completely freeze during this time. It can take upwards of  200+ms for SwapBuffers, GL.Flush, or GL.Finish to complete. I have not tested with SDL backend, due to using GLControl.

It is more apparent on compute shaders with large sizes. But, it can also happen on compute shaders with only DispatchCompute(1,1,1) and no for loops in a simple compute shader where a variable is assigned as output to a Persistent Mapped ShaderStorageBuffer. It seems any time the underlying OpenTK CPU code must sync with the GPU, after a compute shader is used, causes a huge spike in wait time before returning.

### Repro steps

Please provide the steps required to reproduce the problem

* Note: The worker size is set to 8,8,8 in the shader.

1. Use a DispatchCompute(2048 / 8, 2048 / 8, 100 / 8)

2. Use GL.Flush, GL.Finish or SwapBuffers called later. 

### Expected behavior

SwapBuffers, GL.Flush, or GL.Finish  should not be taking 200+ms. Since, the compute shader is already complete by the time these are called.

### Actual behavior

SwapBuffers, GL.Flush or GL.Finish can take upwards of 200ms to complete after a compute shader is launched. Even though the compute shader is finished by the time these are called. This can even happen on small compute shaders that have no for loops and are DispatchCompute(1,1,1);

### Related information 

* Win 10 pro Latest Build
* Nvidia 1080 with Drivers 441.41
* Native Backend
* OpenTK 3.1.0 from Nuget
* .Net Framework 4.6
* No workarounds currently available that I have found.

**Comment by NogginBops** [comment]
This doesn't sound like a OpenTK issue, it sounds like a driver issue to me. Does it happen on other gpus? Are there any driver updates that you can install? Or maybe try the previous version of the driver?

**Comment by Metric** [comment]
This does not happen with any other OpenGL library. Unity 3D OpenGL Core
compute shaders run fine with out having this issue. Unreal 4 OpenGL
Compute shaders run fine as well.

**Comment by jvbsl** [comment]
yeah well high level abstractions don't really tell us anything tbh. What about having a test in C/C++ with OGL. Then it will be clear whether it is an OpenTK issue or not.

**Comment by Metric** [comment]
Sure, I will run a quick test with OGL. Make take a few, I haven’t touched
c++ in a while.

**Comment by Metric** [comment]
Ran two different tests with pure c++ glfw and SDL2. Glfw the swap buffer issue was just as noticeable. SDL2 on the other hand performed as expected with a quick swap.

It seems to be primarily a driver issue as noted prior, but SDL2 did indeed make a noticeable difference, in comparison to glfw.

Tested with a noop compute shader: "#version 430\r\nlayout (local_size_x=8,local_size_y=8,local_size_z=1) in;\r\nvoid main()\r\n{\r\n}";

And with a dispatch compute of:
glDispatchCompute(4096 / 8, 4096 / 8, 100);

Full SDL2 Code:
```
const char* basicComputeSource = "#version 430\r\nlayout (local_size_x=8,local_size_y=8,local_size_z=1) in;\r\nvoid main()\r\n{\r\n}";

int main(int argc, char* argv[]) 
{
	bool running = true;
	SDL_GLContext context;
	SDL_Event sevent;
	SDL_Window* window;

	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4);
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 4);
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);


	if (SDL_Init(SDL_INIT_EVERYTHING) < 0) 
	{
		//failed to init sdl
		return -1;
	}

	if (SDL_Init(SDL_INIT_EVERYTHING) < 0) 
	{
		//failed to init sdl
		return -1;
	}

	window = SDL_CreateWindow("ComputeTest", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 640, 480, SDL_WINDOW_OPENGL | SDL_WINDOW_SHOWN);
	if (window == nullptr)
	{
		//failed to create window
		SDL_Quit();
		return -2;
	}

	context = SDL_GL_CreateContext(window);
	if (context == nullptr)
	{
		//failed to create context
		SDL_DestroyWindow(window);
		window = nullptr;
		SDL_Quit();
		return -3;
	}

	glewExperimental = true;
	glewInit();

	GLuint computeShaderId = 0;
	GLuint computeProgramId = 0;

	while (running) {
		while (SDL_PollEvent(&sevent) != 0) 
		{
			if (sevent.type == SDL_QUIT)
			{
				running = false;
			}
		}

		if (computeProgramId == 0) {

			//create basic compute shader 
			computeShaderId = glCreateShader(GL_COMPUTE_SHADER);
			glShaderSource(computeShaderId, 1, &basicComputeSource, NULL);
			glCompileShader(computeShaderId);
			GLint status;
			glGetShaderiv(computeShaderId, GL_COMPILE_STATUS, &status);
			if (status < 1) {
				//failed to compile shader
				glDeleteShader(computeShaderId);
				SDL_GL_DeleteContext(context);
				SDL_DestroyWindow(window);
				window = nullptr;
				SDL_Quit();
				running = false;
				return -2;
			}

			computeProgramId = glCreateProgram();
			glAttachShader(computeProgramId, computeShaderId);
			glLinkProgram(computeProgramId);
			glGetProgramiv(computeProgramId, GL_LINK_STATUS, &status);
			if (status < 1) {
				//failed to link compute program
				glDetachShader(computeProgramId, computeShaderId);
				glDeleteProgram(computeProgramId);
				glDeleteShader(computeShaderId);
				SDL_GL_DeleteContext(context);
				SDL_DestroyWindow(window);
				window = nullptr;
				SDL_Quit();
				running = false;
				return -3;
			}

			//clean up shader since it is now linked
			glDetachShader(computeProgramId, computeShaderId);
			glDeleteShader(computeShaderId);
		}

		glClearColor(0.2f, 0.3f, 0.3f, 1.0f);
		glClear(GL_COLOR_BUFFER_BIT);

		glUseProgram(computeProgramId);
		glDispatchCompute(4096 / 8, 4096 / 8, 100);

		SDL_GL_SwapWindow(window);
	}

	glDeleteProgram(computeProgramId);

	SDL_GL_DeleteContext(context);
	SDL_DestroyWindow(window);
	window = nullptr;
	SDL_Quit();

	return 0;
}
```

Full glfw code:
```
void framebuffer_size_callback(GLFWwindow* window, int width, int height);
void processInput(GLFWwindow* window);

// settings
const unsigned int SCR_WIDTH = 800;
const unsigned int SCR_HEIGHT = 600;

const char* basicComputeSource = "#version 430\r\nlayout (local_size_x=8,local_size_y=8,local_size_z=1) in;\r\nvoid main()\r\n{\r\n}";

int main(int argc, char *argv[])
{
	// glfw: initialize and configure
	// ------------------------------
	glfwInit();
	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 4);
	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

	// glfw window creation
	// --------------------
	GLFWwindow* window = glfwCreateWindow(SCR_WIDTH, SCR_HEIGHT, "GLComputeTest", NULL, NULL);
	if (window == NULL)
	{
		std::cout << "Failed to create GLFW window" << std::endl;
		glfwTerminate();
		return -1;
	}
	glfwMakeContextCurrent(window);
	glfwSetFramebufferSizeCallback(window, framebuffer_size_callback);

	//init glew
	glewExperimental = true;
	glewInit();

	GLuint computeShaderId = 0;
	GLuint computeProgramId = 0;

	// render loop
	// -----------
	while (!glfwWindowShouldClose(window))
	{
		if (computeProgramId == 0) {

			//create basic compute shader 
			computeShaderId = glCreateShader(GL_COMPUTE_SHADER);
			glShaderSource(computeShaderId, 1, &basicComputeSource, NULL);
			glCompileShader(computeShaderId);
			GLint status;
			glGetShaderiv(computeShaderId, GL_COMPILE_STATUS, &status);
			if (status < 1) {
				//failed to compile shader
				glDeleteShader(computeShaderId);
				glfwTerminate();
				return -2;
			}

			computeProgramId = glCreateProgram();
			glAttachShader(computeProgramId, computeShaderId);
			glLinkProgram(computeProgramId);
			glGetProgramiv(computeProgramId, GL_LINK_STATUS, &status);
			if (status < 1) {
				//failed to link compute program
				glDetachShader(computeProgramId, computeShaderId);
				glDeleteProgram(computeProgramId);
				glDeleteShader(computeShaderId);
				glfwTerminate();
				return -3;
			}

			//clean up shader since it is now linked
			glDetachShader(computeProgramId, computeShaderId);
			glDeleteShader(computeShaderId);
		}

		// input
		// -----
		processInput(window);

		// render
		// ------
		glClearColor(0.2f, 0.3f, 0.3f, 1.0f);
		glClear(GL_COLOR_BUFFER_BIT);

		// glfw: swap buffers and poll IO events (keys pressed/released, mouse moved etc.)
		// -------------------------------------------------------------------------------

		glUseProgram(computeProgramId);
		glDispatchCompute(4096 / 8, 4096 / 8, 100);

		glfwSwapBuffers(window);
		glfwPollEvents();
	}

	glDeleteProgram(computeProgramId);

	// glfw: terminate, clearing all previously allocated GLFW resources.
	// ------------------------------------------------------------------
	glfwTerminate();

	return 0;
}

// process all input: query GLFW whether relevant keys are pressed/released this frame and react accordingly
// ---------------------------------------------------------------------------------------------------------
void processInput(GLFWwindow* window)
{
	if (glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS)
		glfwSetWindowShouldClose(window, true);
}

// glfw: whenever the window size changed (by OS or user resize) this callback function executes
// ---------------------------------------------------------------------------------------------
void framebuffer_size_callback(GLFWwindow* window, int width, int height)
{
	// make sure the viewport matches the new window dimensions; note that width and 
	// height will be significantly larger than specified on retina displays.
	glViewport(0, 0, width, height);
}
```

Will take some time to try different drivers, but this was with the latest Nvidia drivers. In theory though, that 4096 / 8, 4096 / 8, 100 compute should be complete in a couple ms. Especially with a noop. However, glfw had a good stutter of about 10+ms sometimes on swap. Where as SDL2 pretty much performed as you would expect.

**Comment by jvbsl** [comment]
What is the result of SDL_GL_SetSwapInterval and what happens if you set glfwSwapInterval to the same value?

**Comment by Metric** [comment]
Ah I see, setting glfwSwapInterval to 0, did indeed fix the issue in the c++ version. 

However, this leads me to another concern. Why is it not properly being set on the GLControl when I set Vsync to False for both of the GLControls I use in WPF? It is due to creating a shared context between the two controls? Where in, the shared context does not have its Vsync disabled, because  the context is clamped to [1, n)? Rather than [0, n) for SwapInterval?

**Comment by jvbsl** [comment]
Did you make the context current before setting the VSync value?
e.g. like:
```
control1.MakeCurrent();
control1.VSync = false;
control2.MakeCurrent();
control2.VSync = false;
```

**Comment by Metric** [comment]
I was setting the Vsync before the context was ever created, so in theory it should be set to false when the context was created. But, I changed it to the above, and it didn't fix the issue. 

Still experiencing stuttering every so often after a compute shader. Even on the third shared context, for the actual rendering, I am making sure the SwapInterval is set to the lowest allowed, which is 1 currently.

**Comment by Metric** [comment]
I am going to go ahead and close this. It looks like this will be resolved in 4.0. The move to glfw and making the glfw swap interval directly accessible will indeed prevent this issue.
